### PR TITLE
Add functionality to start and shutdown grpc client workers from iovironment

### PR DESCRIPTION
**Pull request by raakella1**

### CODE REVIEW

**Comment by szmyd** [comment]
How does this integrate with nuraft_mesg which doesn't use iomanager?

**Comment by raakella1** [comment]
> How does this integrate with nuraft_mesg which doesn't use iomanager?

Two ways:
1. Do not create any worker threads in nuraft mesg. Let SM estimate the worker threads each of the component needs and create them during startup
2. Let it create its own worker threads and we will rely on ioenvironment to shutdown all workers

**Comment by szmyd** [comment]
> 1. Do not create any worker threads in nuraft mesg. Let SM estimate the worker threads each of the component needs and create them during startup
> 2. Let it create its own worker threads and we will rely on ioenvironment to shutdown all workers

I'm trying to think of a way that nuraft_mesg can present the fact that starting and stopping the client workers is a requirement of the consumer of nuraft_mesg. I don't think it performing *half* the job is a great pattern (as in option 2). So maybe we can just assert in the nuraft_mesg::init code in the event that the workers haven't been started yet? That would make it clear that the responsibility is on the user of the library.

File: src/lib/io_environment.cpp [hunk 1]
```diff
@@ -75,6 +77,11 @@ IOEnvironment& IOEnvironment::with_object_manager() {
     return get_instance();
 }
 
+IOEnvironment& IOEnvironment::with_grpc_client_workers(std::string const& name, uint32_t const num_workers) {
+    sisl::GrpcAsyncClientWorker::create_worker(name, num_workers);
+    return get_instance();
```
Comment (szmyd) [inline src/lib/io_environment.cpp:82]: What are all these `get_instance()` calls? Why not just `*this` like normal? This is a member function, `*this` *is* the singleton...

**Review by szmyd — APPROVED** [review]
lg in general :+1:

**Comment by raakella1** [comment]
> > 1. Do not create any worker threads in nuraft mesg. Let SM estimate the worker threads each of the component needs and create them during startup
> > 2. Let it create its own worker threads and we will rely on ioenvironment to shutdown all workers
> 
> I'm trying to think of a way that nuraft_mesg can present the fact that starting and stopping the client workers is a requirement of the consumer of nuraft_mesg. I don't think it performing _half_ the job is a great pattern (as in option 2). So maybe we can just assert in the nuraft_mesg::init code in the event that the workers haven't been started yet? That would make it clear that the responsibility is on the user of the library.

I like the idea of nuraft_mesg asserting and not doing the half work.

File: src/lib/io_environment.cpp (hunk 1 above)
Comment (raakella1) [reply src/lib/io_environment.cpp:82]: You are right. I started returning get_instance when I first wrote this and am following it for continuity

**Pull request merged**